### PR TITLE
Remove RuntimeClass feature gate and stop serving older versions of RuntimeClass

### PR DESCRIPTION
**Pull request by SergeyKanzhelev**
#### What type of PR is this?

/kind cleanup
/sig node
/priority backlog

#### What this PR does / why we need it:

Feature is GA now

#### Does this PR introduce a user-facing change?
```release-note
The node.k8s.io/v1alpha1 RuntimeClass API is no longer served. Use the node.k8s.io/v1 API version, available since v1.20
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

KEP: https://github.com/kubernetes/enhancements/issues/585

### CODE REVIEW

**Review by ehashman — COMMENTED** [review]
/triage accepted

Did we first need to mark this as deprecated before removal? Looking at https://groups.google.com/g/kubernetes-dev/c/z_AE1EHhZF4/m/fwHC6KSjAwAJ e.g., not sure what the standard path is for API deprecations.

**Comment by SergeyKanzhelev** [comment]
> Did we first need to mark this as deprecated before removal? Looking at https://groups.google.com/g/kubernetes-dev/c/z_AE1EHhZF4/m/fwHC6KSjAwAJ e.g., not sure what the standard path is for API deprecations.

I thought this effectively deprecates the API: https://github.com/kubernetes/kubernetes/pull/103061/files#diff-05cd394e18da9e91dbf8b3d82e1e0813be670ee37db43ca5af6f91a39f78d08f

**Review by ehashman — COMMENTED** [review]
I'm not sure, but someone from API reviewers can probably say for sure :)

/lgtm

**Review by BenTheElder — COMMENTED** [review]
/approve
for vendor/, hack/, test/

**Comment by dchen1107** [comment]
/lgtm
/approve

Can you also announce it at next week's SIG Node? Thanks for the effort.

**Comment by SergeyKanzhelev** [comment]
@kubernetes/api-approvers can you please approve this API removal?

**Comment by liggitt** [comment]
No longer serving this is fine. Dropping the types and conversions in a way that can crash a currently working 1.21 server on upgrade is not fine. Doing this in the release where the v1beta1 version was introduced with a release note that removing existing v1alpha1 runtimeclass instances before upgrade would have been more reasonable, since the only clusters impacted would have been ones that had the alpha API version enabled before upgrade.

If a cluster enabled v1alpha1 runtimeclasses in the past, then upgraded to versions with beta and then GA storage APIs enabled, and never made an update to their persisted runtimeclass objects (which I expect would be common, since these are rarely written), this PR will break the API server's ability to read runtimeclass data from etcd.

**Comment by pacoxu** [comment]
For cleanup of RuntimeClass, 
https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go#L821
RuntimeClass can be removed in this release as well.

**Comment by pacoxu** [comment]
For cleanup of RuntimeClass, 
https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go#L821
RuntimeClass can be removed in this release as well.

**Comment by SergeyKanzhelev** [comment]
> No longer serving this is fine.

I'm not familiar with this part. Where is this configured?

**Comment by liggitt** [comment]
in pkg/registry/node/rest/runtime_class.go: drop v1alpha1Storage / v1beta1Storage

in cmd/kube-apiserver/app/aggregator.go:
```
	{Group: "node.k8s.io", Version: "v1"}:                        {group: 16300, version: 15},
	{Group: "node.k8s.io", Version: "v1alpha1"}:                  {group: 16300, version: 1},
	{Group: "node.k8s.io", Version: "v1beta1"}:                   {group: 16300, version: 9},
```

in pkg/controlplane/instance.go#DefaultAPIResourceConfigSource(), drop nodev1beta1.SchemeGroupVersion and nodev1alpha1.SchemeGroupVersion

since the beta version of runtimeclass is getting dropped in 1.25, it would probably make sense to just drop both alpha and beta then

**Comment by SergeyKanzhelev** [comment]
> in pkg/registry/node/rest/runtime_class.go: drop v1alpha1Storage / v1beta1Storage
> 
> in cmd/kube-apiserver/app/aggregator.go:
> 
> ```
> 	{Group: "node.k8s.io", Version: "v1"}:                        {group: 16300, version: 15},
> 	{Group: "node.k8s.io", Version: "v1alpha1"}:                  {group: 16300, version: 1},
> 	{Group: "node.k8s.io", Version: "v1beta1"}:                   {group: 16300, version: 9},
> ```
> 
> in pkg/controlplane/instance.go#DefaultAPIResourceConfigSource(), drop nodev1beta1.SchemeGroupVersion and nodev1alpha1.SchemeGroupVersion
> 
> since the beta version of runtimeclass is getting dropped in 1.25, it would probably make sense to just drop both alpha and beta then

So I think the test `TestEtcdStoragePath` fails as expected. Would be the best option to modify test to allow "deprecated" types that are not served? Or just abandon the idea?

**Comment by SergeyKanzhelev** [comment]
> /test pull-kubernetes-verify

Just rebased and run `make update` again. No changes made (`Update scripts completed successfully`). Not sure why it's failing. Pushed rebased branch to retest it

File: pkg/controlplane/instance.go [hunk 1]
```diff
@@ -673,7 +670,6 @@ func DefaultAPIResourceConfigSource() *serverstorage.ResourceConfig {
 	// disable alpha versions explicitly so we have a full list of what's possible to serve
 	ret.DisableVersions(
 		apiserverinternalv1alpha1.SchemeGroupVersion,
-		nodev1alpha1.SchemeGroupVersion,
```
Comment (liggitt) [inline pkg/controlplane/instance.go:676]: serving v1alpha1 can be dropped in 1.24

File: pkg/controlplane/instance.go [hunk 2]
```diff
@@ -660,7 +658,6 @@ func DefaultAPIResourceConfigSource() *serverstorage.ResourceConfig {
 		eventsv1beta1.SchemeGroupVersion,
 		networkingapiv1.SchemeGroupVersion,
 		nodev1.SchemeGroupVersion,
-		nodev1beta1.SchemeGroupVersion,
```
Comment (liggitt) [inline pkg/controlplane/instance.go:663]: v1beta1 should be left here until 1.26. RuntimeClass will automatically default to not being served in 1.25, but can be force re-enabled for one release with `KUBE_APISERVER_SERVE_REMOVED_APIS_FOR_ONE_RELEASE=true`

File: pkg/registry/node/rest/runtime_class.go [hunk 3]
```diff
@@ -36,22 +34,6 @@ type RESTStorageProvider struct{}
...
-		}
-	}
-
-	if apiResourceConfigSource.VersionEnabled(nodev1beta1.SchemeGroupVersion) {
-		if storageMap, err := p.v1beta1Storage(apiResourceConfigSource, restOptionsGetter); err != nil {
-			return genericapiserver.APIGroupInfo{}, false, err
-		} else {
-			apiGroupInfo.VersionedResourcesStorageMap[nodev1beta1.SchemeGroupVersion.Version] = storageMap
-		}
-	}
```
Comment (liggitt) [inline pkg/registry/node/rest/runtime_class.go:53]: leave this code in place until 1.26

File: test/integration/etcd/data.go [hunk 4]
```diff
@@ -406,22 +406,6 @@ func GetEtcdStorageDataForNamespace(namespace string) map[schema.GroupVersionRes
...
-		},
-		// --
-
-		// k8s.io/kubernetes/pkg/apis/node/v1beta1
-		gvr("node.k8s.io", "v1beta1", "runtimeclasses"): {
-			Stub:             `{"metadata": {"name": "rc2"}, "handler": "h2"}`,
-			ExpectedEtcdPath: "/registry/runtimeclasses/rc2",
-			ExpectedGVK:      gvkP("node.k8s.io", "v1", "RuntimeClass"),
-		},
-		// --
```
Comment (liggitt) [inline test/integration/etcd/data.go:423]: leave this in place until 1.26

**Review by liggitt — COMMENTED** [review]
removal of the feature gate and v1alpha1 version lgtm (though the release note needs to mention removal of the v1alpha1 version), but the wiring for v1beta1 should be left in place until 1.26

File: pkg/controlplane/instance.go (hunk 2 above)
Comment (liggitt) [reply pkg/controlplane/instance.go:666]: this comment is still outstanding

File: pkg/controlplane/storageversionhashdata/data.go [hunk 5]
```diff
@@ -66,7 +66,6 @@ var GVRToStorageVersionHash = map[string]string{
 	"networking.k8s.io/v1/ingresses":                                   "ZOAfGflaKd0=",
 	"networking.k8s.io/v1/ingressclasses":                              "l/iqIbDgFyQ=",
 	"node.k8s.io/v1/runtimeclasses":                                    "WQTu1GL3T2Q=",
-	"node.k8s.io/v1beta1/runtimeclasses":                               "WQTu1GL3T2Q=",
```
Comment (liggitt) [inline pkg/controlplane/storageversionhashdata/data.go:69]: restore this

File: pkg/registry/node/rest/runtime_class.go (hunk 3 above)
Comment (liggitt) [reply pkg/registry/node/rest/runtime_class.go:53]: still outstanding

File: pkg/registry/node/rest/runtime_class.go [hunk 6]
```diff
@@ -63,28 +45,6 @@ func (p RESTStorageProvider) NewRESTStorage(apiResourceConfigSource serverstorag
...
-func (p RESTStorageProvider) v1beta1Storage(apiResourceConfigSource serverstorage.APIResourceConfigSource, restOptionsGetter generic.RESTOptionsGetter) (map[string]rest.Storage, error) {
-	storage := map[string]rest.Storage{}
-	s, err := runtimeclassstorage.NewREST(restOptionsGetter)
-	if err != nil {
-		return storage, err
-	}
-	storage["runtimeclasses"] = s
-
-	return storage, err
-}
```
Comment (liggitt) [inline pkg/registry/node/rest/runtime_class.go:86]: restore this

File: test/integration/etcd/data.go (hunk 4 above)
Comment (liggitt) [reply test/integration/etcd/data.go:423]: still outstanding

**Comment by liggitt** [comment]
any ETA on the changes requested in https://github.com/kubernetes/kubernetes/pull/103061#pullrequestreview-851872597 ?

File: test/integration/etcd/data.go (hunk 4 above)
Comment (SergeyKanzhelev) [reply test/integration/etcd/data.go]: reverted

**Comment by SergeyKanzhelev** [comment]
> any ETA on the changes requested in [#103061 (review)](https://github.com/kubernetes/kubernetes/pull/103061#pullrequestreview-851872597) ?

should be done now, sorry for the delay.

**Pull request merged**